### PR TITLE
[package] netcdf/4.8.1: Upgrade to 4.8.1.  Fix up patches for 4.7.4.

### PR DESCRIPTION
**Pull request by paulharris**
**Please check the 4.7.4-0001 patch changes I made**, there was a lot of CMAKE_SOURCE_DIR --> PROJECT_SOURCE_DIR changes (including one typo).   I removed those, because I couldn't understand why you would do that (surely there was a reason?).
I removed those changes from the patches because they looked fragile and tricky (original author left a typo mistake), and I haven't seen any other recipe need to worry about such things.  And it would be difficult to maintain.
** **

Otherwise, I manually applied the patches, fixed up patch issues, and compiled the changes into one patch for 4.8.1

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by paulharris** [comment]
I'm not sure why debug-shared builds are failing... any ideas?

**Comment by SpaceIm** [comment]
> Please check the 4.7.4-0001 patch changes I made, there was a lot of CMAKE_SOURCE_DIR --> PROJECT_SOURCE_DIR changes (including one typo)

Because we have a CMakeLists wrapper, but I think these changes were not necessary because they are in dead branches  for this recipe (tests, doc).

**Comment by paulharris** [comment]
The problem was I didn't bring across one of the 4.7.4 patches into 4.8.1

HOWEVER, I feel there is a bug in Conan's cmake finder generator (FindHDF5.cmake)

netcdf key part:
https://github.com/Unidata/netcdf-c/blob/be0cbfdbbeca8ed7d6b44d4cf45fa928a02e3a4e/liblib/CMakeLists.txt#L119

netcdf is trying to use ${HDF5_HL_LIBRARIES},  ${HDF5_C_LIBRARIES},
but HDF5's conan recipe doesn't export any of those.
instead it exports:
- HDF5_LIBS
- HDF5_C_LIBS
- HDF5_LIBRARIES
- but *not* HDF5_C_LIBRARIES

Why doesn't it export package_component_LIBRARIES ?
It seems like existing software expects this from the standard cmake finders.
https://cmake.org/cmake/help/latest/module/FindHDF5.html
```
HDF5_C_LIBRARIES
    Required libraries for the HDF5 C bindings
```

**Comment by paulharris** [comment]
Can I please request some help on this?
The test_package doesn't work (which is based on cmake_find_package),
and it no longer works with my in-progress VTK recipe.
What have I missed?

**Comment by paulharris** [comment]
I think this will work now

**Comment by paulharris** [comment]
I think this SHOULD work.
It should link with `${CURL_LIBRARIES}`, and it works-for-me on my Linux box here.
Can someone please take a look?

**Comment by SpaceIm** [comment]
By moving from `conan_basic_setup()` to `CMakeToolchain`, you have fotgotten something: `conan_basic_setup()` injects interface definitions of all dependencies, which can fix some fragile design in upstream CMakeLists (basically anything not relying on imported targets, like netCDF, usually because they forget that definitions and include dirs are 2 different abstractions in CMake).
So here interface definition of libcurl is not propagated, and it breaks on Windows when libcurl is static...

**Comment by paulharris** [comment]
> Moreover, why do you want to move to CMakeToolchain? It's still unstable.

All the examples generated by "conan new --template cmake_lib" use CMakeToolchain, so naturally as a new conan user I have been using CMT and CMakeDeps ...

Seems to be working "ok" so far, are you saying I should switch to the older generators?

**Comment by SpaceIm** [comment]
I mean that it's premature to refactor conan-center recipes to rely on CMakeToolchain for their own build.
And it's not always as simple as replacing `cmake` generator by `CMakeToolchain` generator, removing `CMakeLists.txt` wrapper, and moving CMake definitions to `generate()`. `conan_basic_setup()` does some magic not available in `CMakeToolchain` (interface definition injected globally as well as folders of libs), so migration of a recipe is not always obvious.

It's doesn't mean that you can't use CMakeToolchain as a consumer of these recipes.

**Comment by paulharris** [comment]
I've changed it back to the older cmake generators, hopefully all is well now

File: recipes/netcdf/all/conanfile.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ class NetcdfConan(ConanFile):
         "with_hdf5": [True, False],
         "cdf5": [True, False],
         "dap": [True, False],
+        "byterange": [True, False], # added to recipe when upgraded to 4.8.1
```
Comment (SpaceIm) [inline recipes/netcdf/all/conanfile.py:27]: this comment can be removed.

File: recipes/netcdf/all/conanfile.py [hunk 2]
```diff
@@ -32,6 +33,7 @@ class NetcdfConan(ConanFile):
         "with_hdf5": True,
         "cdf5": True,
         "dap": True,
+        "byterange": True,
```
Comment (SpaceIm) [inline recipes/netcdf/all/conanfile.py:36]: it's false by default upstream: https://github.com/Unidata/netcdf-c/blob/6a935c68120623a9932cec80e7cad22abe6f6428/CMakeLists.txt#L1053

Comment (paulharris) [reply recipes/netcdf/all/conanfile.py:36]: ok

File: recipes/netcdf/all/conanfile.py (hunk 1 above)
Comment (paulharris) [reply recipes/netcdf/all/conanfile.py:27]: ok

**Comment by paulharris** [comment]
Those changes were applied, as well as adding include/libdirs for the component.

**Review by SSE4 — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Comment by paulharris** [comment]
Thank you for all the attention today!

**Pull request merged**